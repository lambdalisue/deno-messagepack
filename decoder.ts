import { Unpackr } from "https://deno.land/x/msgpackr@v1.8.3/index.js";
import { concat } from "https://deno.land/std@0.184.0/bytes/concat.ts";

const unpackr = new Unpackr({
  useRecords: false,
  structures: [],
});

/** Decode a single MessagePack-encoded object, and returns the decoded object. */
export function decode<T extends unknown>(chunk: ArrayLike<number>): T {
  return unpackr.unpack(chunk) as T;
}

/** Transform a stream of MessagePack-encoded objects into a stream of decoded objects.
 *
 * It support incomplete chunks of MessagePack-encoded objects.
 */
export class DecodeStream<T extends unknown>
  extends TransformStream<Uint8Array, T> {
  #incompleteBuffer?: Uint8Array;

  constructor() {
    super({
      transform: (chunk, controller) => {
        this.#handle(chunk, controller);
      },
    });
  }

  #handle(chunk: Uint8Array, controller: TransformStreamDefaultController<T>) {
    if (this.#incompleteBuffer) {
      chunk = concat(this.#incompleteBuffer, chunk);
      this.#incompleteBuffer = undefined;
    }
    let values: T[] = [];
    try {
      values = unpackr.unpackMultiple(chunk) as T[];
    } catch (err) {
      if (err.incomplete) {
        this.#incompleteBuffer = chunk.slice(err.lastPosition);
        values = err.values ?? [];
      } else {
        throw err;
      }
    } finally {
      for (const value of values) {
        controller.enqueue(value);
      }
    }
  }
}
