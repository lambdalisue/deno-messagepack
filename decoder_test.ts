import { decode, DecodeStream } from "./decoder.ts";
import { assertSnapshot } from "@std/testing/snapshot";

const testcases = {
  string: new Uint8Array([
    163,
    102,
    111,
    111,
  ]),
  number: new Uint8Array([
    100,
  ]),
  array: new Uint8Array([
    150,
    161,
    97,
    0,
    161,
    98,
    1,
    161,
    99,
    2,
  ]),
  object: new Uint8Array([
    222,
    0,
    2,
    163,
    102,
    111,
    111,
    163,
    98,
    97,
    114,
    163,
    98,
    97,
    122,
    163,
    113,
    117,
    120,
  ]),
};

Deno.test("decode", async (t) => {
  for (const [key, value] of Object.entries(testcases)) {
    await t.step(key, async () => {
      const result = decode(value);
      await assertSnapshot(t, result);
    });
  }
});

Deno.test("DecodeStream", async (t) => {
  await t.step("complete", async () => {
    const stream = new ReadableStream({
      start(controller) {
        for (const testcase of Object.values(testcases)) {
          controller.enqueue(testcase);
        }
        controller.close();
      },
    });
    const messages = stream.pipeThrough(new DecodeStream());
    for await (const message of messages) {
      await assertSnapshot(t, message);
    }
  });

  await t.step("incomplete", async () => {
    const stream = new ReadableStream({
      start(controller) {
        for (const testcase of Object.values(testcases)) {
          const head = testcase.slice(0, 4);
          const tail = testcase.slice(4);
          controller.enqueue(head);
          controller.enqueue(tail);
        }
        controller.close();
      },
    });
    const messages = stream.pipeThrough(new DecodeStream());
    for await (const message of messages) {
      await assertSnapshot(t, message);
    }
  });
});
