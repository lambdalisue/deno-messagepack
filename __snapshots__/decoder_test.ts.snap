export const snapshot = {};

snapshot[`decode 1`] = `"foo"`;

snapshot[`decode 2`] = `100`;

snapshot[`decode 3`] = `
[
  "a",
  0,
  "b",
  1,
  "c",
  2,
]
`;

snapshot[`decode 4`] = `
{
  baz: "qux",
  foo: "bar",
}
`;

snapshot[`DecodeStream 1`] = `"foo"`;

snapshot[`DecodeStream 2`] = `100`;

snapshot[`DecodeStream 3`] = `
[
  "a",
  0,
  "b",
  1,
  "c",
  2,
]
`;

snapshot[`DecodeStream 4`] = `
{
  baz: "qux",
  foo: "bar",
}
`;

snapshot[`DecodeStream 5`] = `"foo"`;

snapshot[`DecodeStream 6`] = `100`;

snapshot[`DecodeStream 7`] = `
[
  "a",
  0,
  "b",
  1,
  "c",
  2,
]
`;

snapshot[`DecodeStream 8`] = `
{
  baz: "qux",
  foo: "bar",
}
`;
